### PR TITLE
History page and tests

### PR DESCRIPTION
**Pull request by MaksymNanyzniuk**

### CODE REVIEW

File: web.test.tests/web.test.tests/Features/LoginPage.feature [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿Feature: LoginPage
+	Log in with correct credentials
+
+@mytag @anothetag
```
Comment (varajan) [inline web.test.tests/web.test.tests/Features/LoginPage.feature]: Please review tags, remove unneeded.

File: web.test.tests/web.test.tests/Features/LoginPage.feature [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+Scenario: Success log in
+	Given entered login is "test"
+	And entered password is "newyork1"
+	When click 'Login' button
+	Then Deposit page is opened
+
+
+Scenario: QWERTY
+Given
```
Comment (varajan) [inline web.test.tests/web.test.tests/Features/LoginPage.feature]: Please remove this scenario.

File: web.test.tests/web.test.tests/Pages/DepositPage.cs [hunk 3]
```diff
@@ -71,6 +74,13 @@ public SettingsPage OpenSettingsPage()
             return new SettingsPage(driver);
         }
 
+        public HistoryPage OpenHistoryPage()
+        {
+            History.Click();
+            new WebDriverWait(driver, TimeSpan.FromMilliseconds(10000)).Until(ExpectedConditions.VisibilityOfAllElementsLocatedBy(By.Id(HistoryPage.ClearBtnId)));
+            return new HistoryPage(driver);
```
Comment (varajan) [inline web.test.tests/web.test.tests/Pages/DepositPage.cs]: Are you sure you need this 'wait'? If yes, I suggest move it to a method 'WaitForLoaded' (or something like this) in the HistoryPage class.

```suggestion
            var historyPage = new HistoryPage(driver);
            historyPage.WaitForLoaded();
            return historyPage;
```

File: web.test.tests/web.test.tests/Pages/DepositPage.cs [hunk 4]
```diff
@@ -117,5 +127,38 @@ public void SelectFinYear(int year)
...
+        {
+            CalculateBtn.Submit();
+            Thread.Sleep(500);
+        }
+
+        public Dictionary<String, String> CaptureData()
+        {
+            String year = "";
+            if (FinYear360.GetAttribute("checked") == "true") year = "360";
+            if (FinYear365.GetAttribute("checked") == "true") year = "365";
```
Comment (varajan) [inline web.test.tests/web.test.tests/Pages/DepositPage.cs]: There is 'ternary operator', you may use it in such situations:

```suggestion
            String year = FinYear360.GetAttribute("checked") == "true" ? "365" : "360";
```

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            this.driver = driver;
+        }
+
+        public IWebElement Calculator => driver.FindElement(By.XPath("//div[text()='Calculator']"));
+        public IWebElement ClearBtn => driver.FindElement(By.Id(ClearBtnId));
+
+        public HistoryPage Clear()
+        {
+            ClearBtn.Submit();
+            return new HistoryPage(driver);
```
Comment (varajan) [inline web.test.tests/web.test.tests/Pages/HistoryPage.cs]: If you need to return the same page, you may use 'this'.

```suggestion
            return this;
```

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            try
+            {
+                driver.FindElement(By.XPath($"//table[@id='history']/tr"));
+                return false;
+            } catch (NoSuchElementException e)
+            {
+                return true;
+            }
+        }
```
Comment (varajan) [inline web.test.tests/web.test.tests/Pages/HistoryPage.cs]: FindElement will throw exception, if the element doesn't exist, but FindElement**s** thows no exceptions. So you may write it shorter:

```suggestion
        public bool IsHistoryClean => GetRowsNumber() == 0;
```

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+        public String GetData(int row_number, int column_number)
+        {
+            return GetNRow(row_number).FindElement(By.XPath($"./td[{column_number}]")).Text;
+        }
+
+        public String GetAmount(int row_number)
+        {
+            return GetData(row_number, 1);
+        }
```
Comment (varajan) [inline web.test.tests/web.test.tests/Pages/HistoryPage.cs]: Not wrong, just shorter way:

```suggestion
        public String GetAmount(int row_number) => GetData(row_number, 1);
```

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+        public String GetInterest(int row_number)
+        {
+            return GetData(row_number, 7);
+        }
+
+        public String GetIncome(int row_number)
+        {
+            return GetData(row_number, 8);
+        }
```
Comment (varajan) [inline web.test.tests/web.test.tests/Pages/HistoryPage.cs]: I think mostly all methods above should be 'private', as you use them only inside this class.

File: web.test.tests/web.test.tests/Tests/DepositePageTests.cs [hunk 9]
```diff
@@ -266,5 +254,37 @@ public void Field_Captions_th(String caption)
...
+            DepositPage depositPage = new DepositPage(driver);
+            depositPage.SetAmount(amount);
+            depositPage.SetRate(rate);
+            depositPage.SetTerm(term);
+
+            Assert.IsFalse(depositPage.CalculateBtn.Enabled);
+        }
+
+        [TestCase(1, 1, 1)]
+        public void CalculateBtnEnabled(int amount, int rate, int term)
```
Comment (varajan) [inline web.test.tests/web.test.tests/Tests/DepositePageTests.cs:288]: You don't need this test. If the 'Calculate' button is disabled, the 'Income_And_Interest' test will fail.

File: web.test.tests/web.test.tests/Tests/HistoryPageTests.cs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            //Act
+            SettingsPage settingsPage = depositPage.OpenSettingsPage();
+            settingsPage.SelectDateFormat.SelectByText(dateFormat);
+            settingsPage.Save();
+            HistoryPage historyPage = depositPage.OpenHistoryPage();
+            String histStartDate = historyPage.GetDateFrom(1);
+            String histEndDate = historyPage.GetDateTo(1);
+            
+            //Assert
+            Assert.AreEqual(strStartDate,histStartDate);
```
Comment (varajan) [inline web.test.tests/web.test.tests/Tests/HistoryPageTests.cs]: Please add 'space' for better readability.

```suggestion
            Assert.AreEqual(strStartDate, histStartDate);
```

File: web.test.tests/web.test.tests/Tests/HistoryPageTests.cs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+            Assert.AreEqual(strEndDate, histEndDate);
```
Comment (varajan) [inline web.test.tests/web.test.tests/Tests/HistoryPageTests.cs]: It's better to have one Assert in test, but in this case it's ok to have few Asserts. So it's a good practice to add third parameter with error message.

```suggestion
            Assert.AreEqual(strEndDate, histEndDate, "End Date value is wrong.");
```

**Review by varajan — COMMENTED** [review]
Nice work, I like your way to handle table values!

File: web.test.tests/web.test.tests/Features/LoginPage.feature (hunk 1 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Features/LoginPage.feature]: Removed

File: web.test.tests/web.test.tests/Features/LoginPage.feature (hunk 2 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Features/LoginPage.feature]: Removed

File: web.test.tests/web.test.tests/Pages/DepositPage.cs (hunk 4 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Pages/DepositPage.cs]: Corrected

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs (hunk 5 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Pages/HistoryPage.cs]: Done

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs (hunk 6 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Pages/HistoryPage.cs]: Cool ))

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs (hunk 7 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Pages/HistoryPage.cs]: Done

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs (hunk 8 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Pages/HistoryPage.cs]: Done

File: web.test.tests/web.test.tests/Tests/DepositePageTests.cs (hunk 9 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Tests/DepositePageTests.cs:288]: 'CalculateBtnEnabled' test is created in order to simplify 'Income_And_Interest' test.
 Cases with invalid input data (i.e. when 'Calculate' button is disabled) are not considered in 'Income_And_Interest' test.

File: web.test.tests/web.test.tests/Tests/HistoryPageTests.cs (hunk 10 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Tests/HistoryPageTests.cs]: Done

File: web.test.tests/web.test.tests/Tests/HistoryPageTests.cs (hunk 11 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Tests/HistoryPageTests.cs]: Done

File: web.test.tests/web.test.tests/Pages/DepositPage.cs (hunk 3 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Pages/DepositPage.cs]: No need 'wait' for now.

**Comment by MaksymNanyzniuk** [comment]
Code is corrected according to the comments.

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace web.test.tests.Pages
+{
+    class HistoryPage
+    {
+        private IWebDriver driver;
+
+        public static String ClearBtnId = "clear";
```
Comment (varajan) [inline web.test.tests/web.test.tests/Pages/HistoryPage.cs]: Looks like you don't need this variable. See next comment.

```suggestion
```

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+        public HistoryPage(IWebDriver driver)
+        {
+            this.driver = driver;
+        }
+
+        public IWebElement Calculator => driver.FindElement(By.XPath("//div[text()='Calculator']"));
+        public IWebElement ClearBtn => driver.FindElement(By.Id(ClearBtnId));
```
Comment (varajan) [inline web.test.tests/web.test.tests/Pages/HistoryPage.cs]:
```suggestion
        public IWebElement ClearBtn => driver.FindElement(By.Id("clear"));
```

**Review by varajan — APPROVED** [review]
One minor remark, apply if you like it and let's merge.

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs (hunk 13 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Pages/HistoryPage.cs]: Done

File: web.test.tests/web.test.tests/Pages/HistoryPage.cs (hunk 12 above)
Comment (MaksymNanyzniuk) [reply web.test.tests/web.test.tests/Pages/HistoryPage.cs]: Done

**Pull request merged**